Improvements to build.sh, build documentation and conda developer environments [skip-ci] (#4631)
rapidsai/cuml: BUILD.md

        - [Single and Multi-GPU Components](#single-and-multi-gpu-components)
        - [Configuring Algorithms Built](#configuring-algorithms-built)
    - [Building Python using `setup.py`](#building-python-artifacts-with-setuppy)
- [Unit Tests](#unit-tests)
    - [C++ Unit Tests](#c-unit-tests)
    - [Python Unit Tests](#python-unit-tests)

## Quick Start Guide

1. System must have the following minimum requirements:

- Pascal or newer NVIDIA GPU
- `CUDA` 11.0 or greater. You can obtain CUDA from [https://developer.nvidia.com/cuda-downloads](https://developer.nvidia.com/cuda-downloads).
- `CMake` 3.20.1 or greater
- `GCC`/`G++` 9.3 or greater
- `Ninja` or `make`
- `OpenMP`
- `conda` or `mamba` (`mamba` is recommended to speed up the process.) **Note:** The single GPU C++ artifacts can be built without any further dependencies and without `conda`/`mamba` and the corresponding developer environments, though at the cost of _significantly_ longer compile time and binary size. For more details see [Building C++ Using CMake](#building-c-artifacts-with-cmake).

2. Choose and create the appropriate conda development environment:

- For building the C++ library only:
```bash

> mamba env create -f conda/environments/libcuml_dev_cuda11.5.yml -n libcuml_dev # change libcuml_dev for any name you want for the environment

Reviewer: Nice, I like the separation here.

```

- For building either the C++ or the Python libraries or both:
```bash
mamba env create -f conda/environments/cuml_dev_cuda11.5.yml python=3.9 -n cuml_dev # change libcuml_dev for any name you want for the environment
```

3. From the root folder of the repository, use the convenience script build.sh to build the appropriate artifacts:

- For building the C++ library only:
```bash
PARALLEL_LEVEL=8 ./build.sh libcuml
```

> By default, `Ninja` is used as the cmake generator. To override this and use (e.g.) `make`, define the `CMAKE_GENERATOR` environment variable accodingly:

Reviewer: ```suggestion
By default, `Ninja` is used as the cmake generator. To override this and use (e.g.) `make`, define the `CMAKE_GENERATOR` environment variable accordingly:
```

```bash
PARALLEL_LEVEL=8 CMAKE_GENERATOR='Unix Makefiles' ./build.sh
```

> - For building either the C++ and the Python libraries:

Reviewer: ```suggestion
- For building either the C++ or the Python libraries (or both):
```

```bash
PARALLEL_LEVEL=8 ./build.sh libcuml cuml
```

> In all of the above commands, you can increase the parallel level to speed up compilation with more parallelism, at the cost of needing more RAM in the system.

Reviewer: ```suggestion
In all of the above commands, you can increase the parallel level to speed up compilation with more parallelism at the cost of needing more RAM in the system.
```

4. Run unit tests if desired:

- C++ tests (from the repository root folder):

```bash
cd cpp/build
ninja test
```

- Python tests (from the repository root folder):

```bash
cd python
pytest cuml/test
````

## Setting Up Your Build Environment

> There are two groups of dependencies that need to be met, the core dependencies to build the C++ artifacts, and then the dependencies of the Python package which are a superset of those.

Reviewer: ```suggestion
There are two groups of dependencies that need to be met- the core dependencies to build the C++ artifacts, and then the dependencies of the Python package which are a superset of those.
```

> There are predefined conda environments that meet all the requirements, or they can be added manually. Additionally, many requirements can be fetched automatically by CMake (using CPM), but that can lead to significantly slower build times.

Reviewer: ```suggestion
There are predefined conda environments that can be used to install all the requirements, or they can be added manually. Additionally, many requirements can be fetched automatically by CMake (using CPM), but that can lead to significantly slower build times.
```

```bash
mamba env create -f conda/environments/libcuml_dev_cuda11.5.yml python=3.9 -n libcuml_dev # change libcuml_dev for any name you want for the environment
```

If you already have an environment with the `libcuml++` dependencies, say named `libcuml_dev`, and want to add the Python dependencies, that can be done with:

```bash
mamba env update --file conda/environments/cuml_dev_cuda11.5.yml --name libcuml_dev
```

If a significant amount of time has passed between the creation of your environmente, there might be significant conflicts that could make this update fail, in that case it is suggested to create new environments.

**Note**: If you're using the `rapids_dev_cuda11.5yml` environment that can build all of RAPIDS and want to upgrade any of the packages in it, you must first remove the meta-packages in it with:

```bash
conda remove --force rapids-build-env rapids-notebook-env rapids-doc-env
```

### Docker Developer Container

To use Docker containers for development, there are 2 options, one for building already released RAPIDS versions, and another one that can be used for development verions:

> - For released versions, the containers `rapidsai/rapidsai-dev:22.02-cuda11.5-devel-ubuntu20.04-py3.9` (replacing for the appropiate versions of RAPIDS, Ubuntu and Python) can be used. **Note** This are released as part of the general RAPIDS released version, so they are only apt to build the already released cuML version.

Reviewer: ```suggestion
- For released versions, the containers `rapidsai/rapidsai-dev:22.02-cuda11.5-devel-ubuntu20.04-py3.9` (replacing for the appropiate versions of RAPIDS, Ubuntu and Python) can be used. **Note** These are released as part of the general RAPIDS released version, so they are only apt to build the already released cuML version.
```

9. `libraft-nn` corresponding to the branch/version being built. If not found, it will be fetched by CMake. Using the precompiled binaries from the conda packages speeds up compilation significantly.
10. `treelite`=2.3.0 If not found, it will be fetched by CMake.
11. `libcumlprims` for multiGPU C++ algorithms (Read section on multigpu components).
12. `UCX` with CUDA support >=1.7 for multiGPU C++ algorithms (Read section on multigpu components).
13. `NCCL` (>=2.4) for multiGPU C++ algorithms (Read section on multigpu components).
14. Optional `doxygen` >=1.8.20 for generating documentation

### Python Dependencies

To build the `cuml` Python package, the C++ requirements are needed plus:

15. `cuda-python` corresponding to the CUDA version of the system/environment.
16. `cuDF` corresponding to the branch/version being built (i.e. 22.04 for branch-22.04).
17. `pyraft` corresponding to the branch/version being built (i.e. 22.04 for branch-22.04).
18. `dask-cudf` corresponding to the branch/version being built (i.e. 22.04 for branch-22.04).
19. `dask-cuda` corresponding to the branch/version being built (i.e. 22.04 for branch-22.04).

### Python Unit Test Dependencies

> To run the (`pytest` based) Python unit tests, the following additional packages are needed:

Reviewer: ```suggestion
To run the (`pytest`-based) Python unit tests, the following additional packages are needed:
```

$ pytest cuML/test --collect-only
```

### Full Process

> ### Building C++ Artifacts with CMake

Reviewer: This header seems to be at the same nested level as the one right before it. Was that intentional?

- `"ALL"` is the default option, builds all algorithms into `libcuml++.so`.
- `"CLUSTER"` to include all clustering algorithms, or specify the individual options: `"DBSCAN"`, `"HDBSCAN"`, `"KMEANS"`, `"HIERARCHICALCLUSTERING"`, `"SPECTRALCLUSTERING"`.
- `"DECOMPOSITION"` to include decomposition algorithms, or specify indivual options: `"PCA"`, `"TSVD"`
- `"ENSEMBLE"` or `"RANDOMFOREST"` to include Random Forest (and Decision Tree) algorithms.
- `"FIL"` to include the Forest Inferencing Library (FIL).
- `"KNN"` to include the Nearest Neighbors models.
- `"LINEAR_MODEL"` to include linear models, or specify indivual options: `"LINEARREGRESSION"`, `"RIDGE"`, `"LASSO"`, `"LOGISTICREGRESSION"`.
- `"MANIFOLD"` to include manifold models, or specify indivual options: `"TSNE"`, `"UMAP"`.
- `"METRICS` to include metrics/scoring algorithms.

> - `"SOLVERS"` to include manifold models, or specify indivual options: `"LARS", `"CD", `"SGD", `"QN"

Reviewer: Should this be linear solvers?
